File: examples/java/src/main/java/example/QuickstartTls.java
// ferrokinesis quickstart — AWS SDK for Java (v2) over TLS
//
// Prerequisites:
//   mvn compile
//   ferrokinesis running with TLS:
//     cargo install ferrokinesis --features tls
//     ferrokinesis generate-cert
//     ferrokinesis --tls-cert cert.pem --tls-key key.pem
//
// Usage:
//   mvn compile exec:java -Dexec.mainClass=example.QuickstartTls
//   KINESIS_ENDPOINT=https://localhost:4567 mvn compile exec:java -Dexec.mainClass=example.QuickstartTls

package example;

import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.core.SdkBytes;
import software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.kinesis.KinesisClient;
import software.amazon.awssdk.services.kinesis.model.*;

import javax.net.ssl.*;
import java.net.URI;
import java.security.cert.X509Certificate;

public class QuickstartTls {

    public static void main(String[] args) throws Exception {
        String endpoint = System.getenv().getOrDefault("KINESIS_ENDPOINT", "https://localhost:4567");
        String stream = "java-example-tls";

        // Disable CBOR — ferrokinesis speaks JSON, not CBOR
        System.setProperty("aws.cborEnabled", "false");

        // Trust all certificates (for self-signed certs)
        TrustManager[] trustAll = new TrustManager[]{
                new X509TrustManager() {
                    public X509Certificate[] getAcceptedIssuers() { return new X509Certificate[0]; }
                    public void checkClientTrusted(X509Certificate[] certs, String authType) {}
                    public void checkServerTrusted(X509Certificate[] certs, String authType) {}
                }
        };
        SSLContext sslContext = SSLContext.getInstance("TLS");
        sslContext.init(null, trustAll, new java.security.SecureRandom());

        KinesisClient client = KinesisClient.builder()
                .endpointOverride(URI.create(endpoint))
                .region(Region.US_EAST_1)
                .credentialsProvider(StaticCredentialsProvider.create(
                        AwsBasicCredentials.create("test", "test")))
                .httpClient(UrlConnectionHttpClient.builder()
                        .tlsTrustManagersProvider(() -> trustAll)
                        .build())
                .build();

        // Create a stream
        System.out.println("==> CreateStream");
        client.createStream(CreateStreamRequest.builder()
                .streamName(stream).shardCount(2).build());

        // Wait for ACTIVE
        for (int i = 0; i < 30; i++) {
            DescribeStreamResponse resp = client.describeStream(
                    DescribeStreamRequest.builder().streamName(stream).build());
            if (resp.streamDescription().streamStatus() == StreamStatus.ACTIVE) break;
            Thread.sleep(200);
        }

        // Put a record
        System.out.println("==> PutRecord");
        PutRecordResponse put = client.putRecord(PutRecordRequest.builder()
                .streamName(stream)
                .data(SdkBytes.fromUtf8String("hello world"))
                .partitionKey("pk1")
                .build());

        // Get records
        System.out.println("==> GetRecords");
        GetShardIteratorResponse iterResp = client.getShardIterator(
                GetShardIteratorRequest.builder()
                        .streamName(stream)
                        .shardId(put.shardId())
                        .shardIteratorType(ShardIteratorType.TRIM_HORIZON)
                        .build());

        GetRecordsResponse records = client.getRecords(GetRecordsRequest.builder()
                .shardIterator(iterResp.shardIterator())
                .build());

        for (Record r : records.records()) {
            System.out.println(r.partitionKey() + ": " + r.data().asUtf8String());
        }

        // Clean up
        System.out.println("==> DeleteStream");
        client.deleteStream(DeleteStreamRequest.builder().streamName(stream).build());

        System.out.println("Done.");
    }
}

File: examples/go/quickstart_tls.go
// ferrokinesis quickstart — AWS SDK for Go (v2) over TLS
//
// Prerequisites:
//   go mod tidy
//   ferrokinesis running with TLS:
//     cargo install ferrokinesis --features tls
//     ferrokinesis generate-cert
//     ferrokinesis --tls-cert cert.pem --tls-key key.pem
//
// Usage:
//   go run quickstart_tls.go
//   KINESIS_ENDPOINT=https://localhost:4567 go run quickstart_tls.go

package main

import (
	"context"
	"crypto/tls"
	"fmt"
	"log"
	"net/http"
	"os"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/kinesis"
	"github.com/aws/aws-sdk-go-v2/service/kinesis/types"
)

func main() {
	endpoint := os.Getenv("KINESIS_ENDPOINT")
	if endpoint == "" {
		endpoint = "https://localhost:4567"
	}
	stream := "go-example-tls"

	client := kinesis.New(kinesis.Options{
		Region:       "us-east-1",
		BaseEndpoint: &endpoint,
		Credentials:  credentials.NewStaticCredentialsProvider("test", "test", ""),
		HTTPClient: &http.Client{
			Transport: &http.Transport{
				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
			},
		},
	})
	ctx := context.Background()

	// Create a stream
	fmt.Println("==> CreateStream")
	_, err := client.CreateStream(ctx, &kinesis.CreateStreamInput{
		StreamName: aws.String(stream),
		ShardCount: aws.Int32(2),
	})
	if err != nil {
		log.Fatal(err)
	}

	// Wait for ACTIVE
	for i := 0; i < 30; i++ {
		out, err := client.DescribeStream(ctx, &kinesis.DescribeStreamInput{
			StreamName: aws.String(stream),
		})
		if err == nil && out.StreamDescription.StreamStatus == types.StreamStatusActive {
			break
		}
		time.Sleep(200 * time.Millisecond)
	}

	// Put a record
	fmt.Println("==> PutRecord")
	put, err := client.PutRecord(ctx, &kinesis.PutRecordInput{
		StreamName:   aws.String(stream),
		Data:         []byte("hello world"),
		PartitionKey: aws.String("pk1"),
	})
	if err != nil {
		log.Fatal(err)
	}

	// Get records
	fmt.Println("==> GetRecords")
	iterResp, err := client.GetShardIterator(ctx, &kinesis.GetShardIteratorInput{
		StreamName:        aws.String(stream),
		ShardId:           put.ShardId,
		ShardIteratorType: types.ShardIteratorTypeTrimHorizon,
	})
	if err != nil {
		log.Fatal(err)
	}

	records, err := client.GetRecords(ctx, &kinesis.GetRecordsInput{
		ShardIterator: iterResp.ShardIterator,
	})
	if err != nil {
		log.Fatal(err)
	}

	for _, r := range records.Records {
		fmt.Printf("%s: %s\n", *r.PartitionKey, string(r.Data))
	}

	// Clean up
	fmt.Println("==> DeleteStream")
	_, err = client.DeleteStream(ctx, &kinesis.DeleteStreamInput{
		StreamName: aws.String(stream),
	})
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("Done.")
}
